### PR TITLE
Curve lattice twist optate

### PR DESCRIPTION
**Pull request by plabanca**

### CODE REVIEW

**Comment by spoonincode** [comment]
bn256_test.cpp appears to need an `#include <memory>` otherwise std::unique_ptr isn't found when using libstdc++ 12 (gcc12)

File: src/curve.cpp [hunk 1]
```diff
@@ -0,0 +1,221 @@
+#include <curve.h>
+#include <gfp_generic.h>
+#include <constants.h>
+#include <lattice.h>
+
+using boost::multiprecision::int512_t;
+namespace bn256 {
+
+    static constexpr gfp curve_b {3};
```
Comment (linh2931) [inline src/curve.cpp]: Our code base uses 3 spaces for indentation. Please conform to it. This comment applies to all the files.

lattice.h has very long spaces (9) in some statements.

File: src/lattice.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    using namespace boost::multiprecision::literals;
+    using namespace boost::multiprecision;
+
+    // decompose takes a scalar mod order as input and finds a short,
+    // positive decomposition of it wrt to the lattice basis.
+    std::vector<int512_t> lattice::decompose(const int512_t& k) const {
+        auto n = inverse_.size();
+        std::vector<int512_t> c(n);
+
```
Comment (linh2931) [inline src/lattice.cpp:14]: Add the missing comment `// Calculate closest vector in lattice to <k,0,0,...> with Babai's rounding.`

File: src/lattice.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            c[i] = k * inverse_[i];
+            round(c[i], det_);
+        }
+
+        // Transform vectors according to c and subtract <k,0,0,...>.
+        std::vector<int512_t> out(n);
+        int512_t temp;
+
+        for (std::size_t i = 0; i < n; i++) {
+
```
Comment (linh2931) [inline src/lattice.cpp:40]: Remove empty line

File: src/lattice.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            out[i] = 0;
+
+            for (std::size_t j = 0; j < n; j++) {
+                temp = c[j] * vectors_[j][i];
+                out[i] = out[i] + temp;
```
Comment (linh2931) [inline src/lattice.cpp]: Change to `out[i] += temp;`

File: src/lattice.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+            const auto len = bitlen(x);
+            if (len > maxlen) {
+                maxlen = len;
+            }
+        }
+
+        std::vector<uint8_t> out(maxlen);
+        for (auto& x : decomp) {
+            const auto j = &x - &decomp[0];
```
Comment (linh2931) [inline src/lattice.cpp]: Can` j` use explicit value rather than address manipluation?

File: src/optate.cpp [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+        rc.out_.x_.sub(rc.out_.x_, v);
+
+        rc.out_.z_.add(rc.out_.z_, h);
+        rc.out_.z_.square(rc.out_.z_);
+        rc.out_.z_.sub(rc.out_.z_, r.t_);
+        rc.out_.z_.sub(rc.out_.z_, i);
+
+        t.sub(v, rc.out_.x_);
+        t.mul(t, l1);
+        t2.mul(rc.out_.y_, j);
```
Comment (linh2931) [inline src/optate.cpp]: Should `rc.out_.y_` be `r.y_`?

File: src/optate.cpp [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        rc.out_.z_.sub(rc.out_.z_, r.t_);
+
+        rc.out_.y_.sub(d, rc.out_.x_);
+        rc.out_.y_.mul(rc.out_.y_, e);
+
+        t.add(c, c);
+        t.add(t, t);
+        t.add(t, t);
+
+        rc.out_.y_.sub(rc.out_.y_, t);
```
Comment (linh2931) [inline src/optate.cpp]: I can't find where those came from.

File: src/optate.cpp [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+    gfp12 optimal_ate(const twist_point& a, const curve_point& b) {
+        auto e = miller(a, b);
+        auto ret = final_exponentiation(e);
+
+        if (a.is_infinity() || b.is_infinity()) {
+            ret.set_one();
+        }
+        return ret;
+    }
+
```
Comment (linh2931) [inline src/optate.cpp]: remove empty line

File: src/curve.h [hunk 9]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+#include <gfp.h>
+#include <boost/multiprecision/cpp_int.hpp>
+
+using boost::multiprecision::int512_t;
+
+namespace bn256 {
+
```
Comment (linh2931) [inline src/curve.h:9]: Add the comments from the original file.

File: src/curve.cpp [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+        // This is also
+        // y_ = - 2·s1·j - (s2-s1)(2x - 2i·u1) = r(v-x_) - 2·s1·j
+        gfp_sub(t, v, x_);   // t7
+        gfp_mul(t4, s1, j);  // t8
+        gfp_add(t6, t4, t4); // t9
+        gfp_mul(t4, r, t);   // t10
+        gfp_sub(y_, t4, t6);
+
+        // Set z_ = 2(u2-u1)·z1·z2 = 2h·z1·z2
+        gfp_add(t, a.z_, b.z_);       // t11
```
Comment (linh2931) [inline src/curve.cpp]: Align the comment

File: src/curve.cpp [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+
+    void curve_point::make_affine() {
+        if (z_ == new_gfp(1ll)) {
+            return;
+        } else if (z_ == new_gfp(0ll)) {
+            x_.fill(0);
+            y_ = new_gfp(1ll);
+            t_.fill(0);
+            return;
+        }
```
Comment (linh2931) [inline src/curve.cpp]: Why the following code from the original was missing?

 ```
 zInv := &gfP{}
  zInv.Invert(&c.z)

  t, zInv2 := &gfP{}, &gfP{}
  gfpMul(t, &c.y, zInv)
  gfpMul(zInv2, zInv, zInv)

  gfpMul(&c.x, &c.x, zInv2)
  gfpMul(&c.y, t, zInv2)

  c.z = *newGFp(1)
  c.t = *newGFp(1)
```

File: src/curve.cpp [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    std::string curve_point::string() {
+        make_affine();
+
+        gfp x{}, y{};
+        x.mont_decode(x_);
+        y.mont_decode(y_);
+
+        std::stringstream ss;
```
Comment (linh2931) [inline src/curve.cpp]: You can just use std::string for ss and use std::string's operations.

File: src/curve.cpp [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    }
+
+    void curve_point::neg(const curve_point& a) {
+        x_.set(a.x_);
+        gfp_neg(y_, a.y_);
+        z_.set(a.z_);
+        t_.set(a.t_);
+    }
+
+    std::string curve_point::string() {
```
Comment (linh2931) [inline src/curve.cpp]: Why was not this function placed in the beginning as the original one?

File: src/twist.h [hunk 14]
```diff
@@ -0,0 +1,36 @@
+#pragma once
+#include <gfp2.h>
+#include <boost/multiprecision/cpp_int.hpp>
+
+using boost::multiprecision::int512_t;
+
```
Comment (linh2931) [inline src/twist.h:7]: Add the comments from the original.

File: src/twist.cpp [hunk 15]
```diff
@@ -0,0 +1,194 @@
...
+            set(b);
+            return;
+        }
+        if (b.is_infinity()) {
+            set(a);
+            return;
+        }
+
+        // See http://hyperelliptic.org/EFD/g1p/auto-code/shortw/jacobian-0/addition/add-2007-bl.op3
+        gfp2 z12{}, z22{}, u1{}, u2{}, t{}, s1{}, s2{}, h{}, i{}, j{}, r{}, v{}, t4{}, t6{};
```
Comment (linh2931) [inline src/twist.cpp]: It would be better to declare variables next to their first uses.

File: src/twist.cpp [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+        x_.sub(t6, t);
+
+        t.sub(v, x_);   // t7
+        t4.mul(s1, j);  // t8
+        t6.add(t4, t4); // t9
+        t4.mul(r, t);   // t10
+        y_.sub(t4, t6);
+
+        t.add(z_, b.z_); // t11
+        t4.square(t);      // t12
```
Comment (linh2931) [inline src/twist.cpp]: align the comment

File: src/twist.cpp [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+        t.sub(t4, z12);  // t13
+        t4.sub(t, z22);  // t14
+        z_.mul(t4, h);
+
+    }
+
+    void twist_point::double_(const twist_point& a) {
+        // See http://hyperelliptic.org/EFD/g1p/auto-code/shortw/jacobian-0/doubling/dbl-2009-l.op3
+        gfp2 a2{}, b{}, c{}, t{}, t2{}, d{}, e{}, f{};
```
Comment (linh2931) [inline src/twist.cpp]: It might be better keep the cases  of the original so you don't need to use `a2`

File: src/twist.cpp [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+        t6.add(t4, t4); // t9
+        t4.mul(r, t);   // t10
+        y_.sub(t4, t6);
+
+        t.add(z_, b.z_); // t11
+        t4.square(t);      // t12
+        t.sub(t4, z12);  // t13
+        t4.sub(t, z22);  // t14
+        z_.mul(t4, h);
+
```
Comment (linh2931) [inline src/twist.cpp]: Remove empty line. This applies to other functions too.

File: src/twist.cpp [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    void twist_point::neg(const twist_point& a) {
+        x_.set(a.x_);
+        y_.neg(a.y_);
+        z_.set(a.z_);
+        t_.set_zero();
+    }
+
+    std::string twist_point::string() {
```
Comment (linh2931) [inline src/twist.cpp]: The comment of curve.cpp applies to here.

**Review by linh2931 — CHANGES_REQUESTED** [review]

File: src/curve.cpp (hunk 1 above)
Comment (plabanca) [reply src/curve.cpp]: Resolved with commit: https://github.com/AntelopeIO/bn256/pull/2/commits/43addb83e26bc0f0a281f4b451354925da063af3

File: src/optate.cpp (hunk 6 above)
Comment (plabanca) [reply src/optate.cpp]: Yes, it should:

	t := (&gfP2{}).Sub(V, &rOut.x)
	t.Mul(t, L1)
	**t2 := (&gfP2{}).Mul(&r.y, J)**

File: src/optate.cpp (hunk 7 above)
Comment (plabanca) [reply src/optate.cpp]: **optate.go: lines 73-74:**

rOut.y.Sub(D, &rOut.x).Mul(&rOut.y, E)
**t := (&gfP2{}).Add(C, C)
t.Add(t, t).Add(t, t)**
rOut.y.Sub(&rOut.y, t)

Comment (linh2931) [reply src/optate.cpp]: I must have misread.

**Review by linh2931 — APPROVED** [review]

**Pull request merged**